### PR TITLE
[ACR] Just accept digest in delete_blob()

### PR DESCRIPTION
**Pull request by YalinLi0312**

### CODE REVIEW

**Comment by YalinLi0312** [comment]
Don't have test updates in this PR, because 1. sync test already have usage of delete_blob() 2. async upload_blob() and download_blob() implementation is still under review in other PRs, the async delete test will be covered there.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 1]
```diff
@@ -1037,9 +1036,6 @@ def delete_blob(self, repository: str, tag_or_digest: str, **kwargs) -> None:
...
             endpoint = os.environ["CONTAINERREGISTRY_ENDPOINT"]
             client = ContainerRegistryClient(endpoint, DefaultAzureCredential(), audience="my_audience")
-            client.delete_blob("my_repository", "my_tag_or_digest")
+            client.delete_blob("my_repository", "my_digest")
         """
-        if _is_tag(tag_or_digest):
-            tag_or_digest = self._get_digest_from_tag(repository, tag_or_digest)
-
-        self._client.container_registry_blob.delete_blob(repository, tag_or_digest, **kwargs)
+        self._client.container_registry_blob.delete_blob(repository, digest, **kwargs)
```
Comment (annelo-msft) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:1041]: This is now consistent with what other languages are doing; however, is it a breaking change to your library?

Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:1041]: No, this operation hasn't GAed. (The GAed one is `delete_manifest`)

Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:1041]: Well, it's actually a breaking change with our last preview. So it should be acceptable.

Comment (annelo-msft) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:1041]: Got it - thanks for the clarification!

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py [hunk 2]
```diff
@@ -864,3 +864,24 @@ async def delete_manifest(self, repository: str, tag_or_digest: str, **kwargs) -
             tag_or_digest = await self._get_digest_from_tag(repository, tag_or_digest)
 
         await self._client.container_registry.delete_manifest(repository, tag_or_digest, **kwargs)
+
+    @distributed_trace_async
+    async def delete_blob(self, repository: str, digest: str, **kwargs) -> None:
+        """Delete a blob. If the blob cannot be found or a response status code of
+        404 is returned an error will not be raised.
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py:871]: It says here that an error will not be raised in the case of a 404, but I don't see any handling here of a 404. The generated code will raise on a 404. So does the service never return a 404? Or should we be handling it here?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 1 above)
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:1041]: Same comment as async - the docstring implies that we're handling the `ResourceNotFoundError` - but I don't see that here.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py (hunk 2 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py:871]: Service does return a 404 error `ResourceNotFoundError` when blob to delete is not found, we are not handling the error here. The docstring here is not accurate.

**Comment by YalinLi0312** [comment]
Added tests for 404 in sync and async.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 3]
```diff
@@ -1037,9 +1036,10 @@ def delete_blob(self, repository: str, tag_or_digest: str, **kwargs) -> None:
...
         """
-        if _is_tag(tag_or_digest):
-            tag_or_digest = self._get_digest_from_tag(repository, tag_or_digest)
-
-        self._client.container_registry_blob.delete_blob(repository, tag_or_digest, **kwargs)
+        try:
+            self._client.container_registry_blob.delete_blob(repository, digest, **kwargs)
+        except HttpResponseError as error:
+            if error.status_code == 404:
+                return
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:1045]: You need to add the final `raise` or all errors will be suppressed.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py [hunk 4]
```diff
@@ -864,3 +864,28 @@ async def delete_manifest(self, repository: str, tag_or_digest: str, **kwargs) -
...
+            from azure.identity.aio import DefaultAzureCredential
+            endpoint = os.environ["CONTAINERREGISTRY_ENDPOINT"]
+            client = ContainerRegistryClient(endpoint, DefaultAzureCredential(), audience="my_audience")
+            await client.delete_blob("my_repository", "my_digest")
+        """
+        try:
+            await self._client.container_registry_blob.delete_blob(repository, digest, **kwargs)
+        except HttpResponseError as error:
+            if error.status_code == 404:
+                return
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_container_registry_client.py:891]: Also needs final `raise`

**Review by annatisch — APPROVED** [review]

**Pull request merged**